spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/PulsarAutoConfiguration.java

```diff
@@ -78,21 +77,31 @@ public class PulsarAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean(PulsarProducerFactory.class)
-	DefaultPulsarProducerFactory<Object> pulsarProducerFactory(Environment environment, PulsarClient pulsarClient,
-			TopicResolver topicResolver, ObjectProvider<ProducerBuilderCustomizer<?>> customizersProvider) {
-		String topicName = this.properties.getProducer().getTopicName();
+	@ConditionalOnProperty(name = "spring.pulsar.producer.cache.enabled", havingValue = "false")
+	DefaultPulsarProducerFactory<?> pulsarProducerFactory(PulsarClient pulsarClient, TopicResolver topicResolver,
+			ObjectProvider<ProducerBuilderCustomizer<?>> customizersProvider) {
+		return new DefaultPulsarProducerFactory<>(pulsarClient, this.properties.getProducer().getTopicName(),
+				lambdaSafeProducerBuilderCustomizers(customizersProvider), topicResolver);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(PulsarProducerFactory.class)
+	@ConditionalOnProperty(name = "spring.pulsar.producer.cache.enabled", havingValue = "true", matchIfMissing = true)
+	CachingPulsarProducerFactory<?> cachingPulsarProducerFactory(PulsarClient pulsarClient, TopicResolver topicResolver,
+			ObjectProvider<ProducerBuilderCustomizer<?>> customizersProvider) {
+		return new CachingPulsarProducerFactory<>(pulsarClient, this.properties.getProducer().getTopicName(),
+				lambdaSafeProducerBuilderCustomizers(customizersProvider), topicResolver,
+				this.properties.getProducer().getCache().getExpireAfterAccess(),
+				this.properties.getProducer().getCache().getMaximumSize(),
+				this.properties.getProducer().getCache().getInitialCapacity());
+	}
+
+	private List<ProducerBuilderCustomizer<Object>> lambdaSafeProducerBuilderCustomizers(
+			ObjectProvider<ProducerBuilderCustomizer<?>> customizersProvider) {
 		List<ProducerBuilderCustomizer<?>> customizers = new ArrayList<>();
 		customizers.add(PulsarPropertyMapper.producerBuilderCustomizer(this.properties));
 		customizers.addAll(customizersProvider.orderedStream().toList());
-		List<ProducerBuilderCustomizer<Object>> lambdaSafeCustomizers = List
```

Comment:
@philwebb I broke this out into 2 separate beans for consistency w/ the reactive side and to have more concrete bean return types.